### PR TITLE
Added ProcessingTimeMetric

### PR DESCRIPTION
**Pull request by indualagarsamy**
Relates to https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/49

Reporting is not done yet.

cc / @mikeminutillo @SzymonPobiega

### CODE REVIEW

File: src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        /// This enables reporting of the ProcessingTime metric. To disable this feature, call: endpointConfig.DisableFeature&lt;ProcessingTimeMetric&gt;() 
+        /// </summary>
+        /// <param name="context">feature context</param>
+        protected override void Setup(FeatureConfigurationContext context)
+        {
+            context.ThrowIfSendonly();
+            context.Pipeline.OnReceivePipelineCompleted(e =>
+            {
+                DateTime timeSent;
+                if (e.TryGetTimeSent(out timeSent))
```
Comment (mikeminutillo) [inline src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs]: The `timeSent` value is never used. It probably will be if we implement Critical Time here.

File: src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                {
+                    string messageTypeProcessed;
+                    e.TryGetMessageHandlerType(out messageTypeProcessed);
+                    reporter.Report(timeSent, messageTypeProcessed, e.StartedAt, e.CompletedAt);
+                }
+                return Task.FromResult(0);
```
Comment (mikeminutillo) [inline src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs]: Do we have access to `Task.CompletedTask` here?

File: src/NServiceBus.Monitoring/ProcessingTime/ReportProcessingTime.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    using System;
+    using System.Diagnostics;
+
+    class ReportProcessingTime
+    {
+        public void Report(DateTime timeSent, string processedMessageType,  DateTime startedAt, DateTime completedAt)
+        {
+            var computedTimeElapsed = ProcessingTimeCalculator.Calculate(startedAt, completedAt);
+
+            var totalTimeInSeconds = Convert.ToInt32(computedTimeElapsed);
```
Comment (mikeminutillo) [inline src/NServiceBus.Monitoring/ProcessingTime/ReportProcessingTime.cs]: Should this time be measured in seconds or milliseconds?

`TimeSpan` has a `TotalSeconds` (and a `TotalMilliseconds`) property on it.

Comment (indualagarsamy) [inline src/NServiceBus.Monitoring/ProcessingTime/ReportProcessingTime.cs]: @mikeminutillo - I think, milliseconds make more sense. Especially if there are tools that try to graph it, etc and need a more granular measurement. We can always round to seconds, but we can't get the original value back.

File: src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs (hunk 1 above)
Comment (indualagarsamy) [inline src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs]: @mikeminutillo - yeah, although not used at the moment, like you said can be used for CriticalTime

File: src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs (hunk 2 above)
Comment (indualagarsamy) [inline src/NServiceBus.Monitoring/ProcessingTime/ProcessingTimeMetric.cs]: Not if we are using .NET 4.5.2. Task.CompletedTask - isn't that only available in .Net 4.6.x?

**Comment by indualagarsamy** [comment]
@mikeminutillo - regarding the reliance on 6.2, Perfcounters package also has the same limitation. I think for a person using Version 6 and 6.2, setting the min dependency to 6.2 shouldn't be a big deal. If on the other hand, there were a lot of changes that people need to do to convert 6.0 to 6.2, then that might be a blocker which I think not.

**Pull request merged**